docs/proposals/cluster-autoscaling.md

```diff
@@ -0,0 +1,324 @@
+* (I6) [Hard to enable/disable] user can easily update the configuration on a running cluster.
+
+It will also allow us to switch the implementation later without changing the interface (too much).
+
+This will not solve the following problems:
+
+(I2) [GCE specific logic] To port this solution to other cloud providers one would have to reimplement whole logic that
+implements common interface (or configure cloud-provider specific autoscaler); it’s possible but hard.
+(I3) [Works only for well formed pods] We will also introduce an experimental/hacky metrics for pending pods that will be useful
+for scaling up clusters when the pods are larger than 1-target. This will be only a help, not a full solution.
+
+For development/experimentation/proof-of-concept KAC config it will be stored in Etcd as an API object. We will consider 
+other storage options once they become available.
+
+```go
+
+type NodeResource string
+
+const (
+    // Percentage of node's CPU that is currently used.
+    CPUConsumption NodeResource = "CPUConsumption"
+
+    // Percentage of node's CPU that is currently requested for pods.
+    CPURequest NodeResource = "CPURequest"
+
+    // Percentage of node's memory that is currently used.
+    MemoryConsumption NodeResource = "MemoryConsumption"
+
+    // Percentage of node's memory that is currently requested for pods.
+    MemoryRequest NodeResource = "MemoryRequest"
+)
+
+// NodeUtilization describes what percentage of a particular resource is used or requested on a node.
+type NodeUtilization struct {
+    Resource NodeResource `json:"resource"`
+
+    // The accepted values are from 0 to 100.
+    Percentage int `json:"percentage"`
+}
+
+// Configuration of the Cluster Autoscaler
+type ClusterAutoscalerSpec struct {
+    // Minimum number of nodes that the cluster should have.
+    MinNodes int `json:"minNodes"`
+
+    // Maximum number of nodes that the cluster should have.
+    MaxNodes int `json:"maxNodes"`
+
+    // Target average utilization of the cluster nodes. New nodes will be added if one (or more) of the
+    // targets is exceeded. Cluster size will be decreased if the current utilization is too low
+    // for all targets.
+    TargetUtilization []NodeUtilization `json:"targetUtilization"`
+}
+
+type ClusterAutoscaler struct {
+    // To be deleted once the configuration is moved out of APIServer.
+    unversioned.TypeMeta `json:",inline"`
+
+    // Standard object's metadata.
+    // More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
+    // For now (experimental api) it is required that the name is set to "ClusterAutoscaler" and namespace is "".
+    v1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec defines the desired behavior of this ClusterAutoscaler.
+    // More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status
+    Spec ClusterAutoscalerSpec `json:"spec,omitempty"`
```

Comment:
Should there be any kind of status, for example the current size of the cluster?